pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BrokerRequestHandlerDelegate.java

```diff
@@ -0,0 +1,113 @@
+ * under the License.
+ */
+package org.apache.pinot.broker.requesthandler;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import javax.annotation.Nullable;
+import org.apache.pinot.broker.api.RequesterIdentity;
+import org.apache.pinot.common.response.BrokerResponse;
+import org.apache.pinot.spi.trace.RequestContext;
+import org.apache.pinot.spi.utils.CommonConstants;
+
+
+/**
+ * {@code BrokerRequestHandlerDelegate} delegates the inbound broker request to one of the enabled
+ * {@link BrokerRequestHandler} based on the requested handle type.
+ *
+ * {@see: @CommonConstant
+ */
+public class BrokerRequestHandlerDelegate implements BrokerRequestHandler {
+
+  private String _defaultBrokerRequestHandlerType;
+  @Nullable
+  private SingleConnectionBrokerRequestHandler _singleConnectionBrokerRequestHandler;
+  @Nullable
+  private GrpcBrokerRequestHandler _grpcBrokerRequestHandler;
+  @Nullable
+  private WorkerQueryRequestHandler _multiStageWorkerRequestHandler;
+
+  public BrokerRequestHandlerDelegate(@Nullable String defaultBrokerRequestHandlerType,
+      @Nullable SingleConnectionBrokerRequestHandler singleConnectionBrokerRequestHandler,
+      @Nullable GrpcBrokerRequestHandler grpcBrokerRequestHandler,
+      @Nullable WorkerQueryRequestHandler multiStageWorkerRequestHandler
+  ) {
+    _defaultBrokerRequestHandlerType = defaultBrokerRequestHandlerType != null ? defaultBrokerRequestHandlerType
+        : CommonConstants.Broker.DEFAULT_BROKER_REQUEST_HANDLER_TYPE;
+    _singleConnectionBrokerRequestHandler = singleConnectionBrokerRequestHandler;
+    _grpcBrokerRequestHandler = grpcBrokerRequestHandler;
+    _multiStageWorkerRequestHandler = multiStageWorkerRequestHandler;
+  }
+
+  @Override
+  public void start() {
+    if (_singleConnectionBrokerRequestHandler != null) {
+      _singleConnectionBrokerRequestHandler.start();
+    }
+    if (_grpcBrokerRequestHandler != null) {
+      _grpcBrokerRequestHandler.start();
+    }
+    if (_multiStageWorkerRequestHandler != null) {
+      _multiStageWorkerRequestHandler.start();
+    }
+  }
+
+  @Override
+  public void shutDown() {
+    if (_singleConnectionBrokerRequestHandler != null) {
+      _singleConnectionBrokerRequestHandler.shutDown();
+    }
+    if (_grpcBrokerRequestHandler != null) {
+      _grpcBrokerRequestHandler.shutDown();
+    }
+    if (_multiStageWorkerRequestHandler != null) {
+      _multiStageWorkerRequestHandler.shutDown();
+    }
+  }
+
+  @Override
+  public BrokerResponse handleRequest(JsonNode request, @Nullable RequesterIdentity requesterIdentity,
+      RequestContext requestContext)
+      throws Exception {
+    JsonNode node = request.get(CommonConstants.Broker.BROKER_REQUEST_HANDLER_TYPE_JSON_OVERRIDE_KEY);
+    String handlerType = node == null ? _defaultBrokerRequestHandlerType : node.asText();
+    switch (handlerType) {
+      case CommonConstants.Broker.NETTY_BROKER_REQUEST_HANDLER_TYPE:
+        if (_singleConnectionBrokerRequestHandler != null) {
```

Comment:
One concern is that this can lead to potential overhead for very high throughput (e.g in our case 30000) as the if check will be done for every query (ignoring any compiler optimizations for now).

Can we consider providing a fast path or something that is on parity for the queries going down the existing request handlers ? For example, anyone that does not enable the multi-stage query engine will still have this potential penalty due to if check upon every request

Reply:
let me think about how to do a lazy singleton default instantiation, when there's no mulit-stage query engine enable flag.

Reply:
lazy default implemented